source/AI.cpp

```diff
@@ -103,6 +103,109 @@ namespace {
 		return false;
 	}
 	
+	// Determine if the ship with the given travel plan should refuel in
+	// its current system, or if it should keep travelling.
+	bool ShouldRefuel(const Ship &ship, const DistanceMap &route)
+	{
+		double fuelCapacity = ship.Attributes().Get("fuel capacity");
+		const bool hasFuel = fuelCapacity && ship.JumpFuel();
+		
+		const System *from = ship.GetSystem();
+		const bool systemHasFuel = from->HasFuelFor(ship);
+		
+		// Calculate the fuel needed to reach the next system with fuel.
+		const System *to = route.Route(from);
+		if(systemHasFuel && hasFuel && ship.Fuel() < 1.)
+		{
+			while(to && !to->HasFuelFor(ship))
+				to = route.Route(to);
+			
+			// If the ship cannot route further (!to), or needs more
+			// fuel to reach the next system with fuel, refuel.
+			if(!to || ship.Fuel() * fuelCapacity < route.RequiredFuel(from, to))
+				return true;
+			else
+				return false;
+		}
+		return false;
+	}
+	// Wrapper for ship-system uses.
+	bool ShouldRefuel(const Ship &ship, const System *to)
+	{
+		if(!to || ship.Fuel() == 1. || !ship.GetSystem()->HasFuelFor(ship))
+			return false;
+		double needed = ship.JumpFuel(to);
+		if(needed && to->HasFuelFor(ship))
+			return ship.Fuel() * ship.Attributes().Get("fuel capacity") < needed;
```

Comment:
When there is no fuel capacity, it tells the ship to refuel.

Reply:
Should be addressed now.